### PR TITLE
Task/custom module calls

### LINKED ISSUE #37: Create a module

### PR DESCRIPTION
**Pull request by venoivankovic**
* Added rbac module
* Added API routes to call deployed rbac module, deployed module can be found here https://goerli.etherscan.io/address/0xfc3c71d00d522f8a2537523b22f54405011257ff
* Currently using this safe https://app.safe.global/transactions/history?safe=gor:0x1f29deFa8a6472f65e1c86cD8E449cd37c521A41
* Added scripts to enable and disable a module on a Safe, enable-module needs to be run to add the plugin module to the safe
* Safe owners can call add-user route to add a user to the safe, remove-user to remove the user from the safe
* The user can then call the user-transaction route to make a transaction from the Safe

### CODE REVIEW

File: contracts/rbac-module/contracts/Enum.sol
Comment (venoivankovic) [inline contracts/rbac-module/contracts/Enum.sol:1]: the contracts folder should not look like this at all
there should be a deployment script and unit tests, similar to how hardhat projects look (in fact this dir should be a hardhat project imo)
i think its best to do that in a separate pr

File: apps/Wallet/src/types/safe-transaction.ts [hunk 1]
```diff
@@ -11,3 +11,21 @@ export type FundUserTransactionBody = {
 export type AddOwnerTransactionBody = {
   ownerAddress: string;
 }
+
+export type AddUserTransactionBody = {
```
Comment (venoivankovic) [inline apps/Wallet/src/types/safe-transaction.ts]: should these be just one type, UserTransactionBody, or better like this?

**Comment by GabrielKidane96** [comment]
Instead of ```const moduleContract = new ethers.Contract(allowanceModuleAddress, rbac, signer) as Rbac;```
Use typechain as intended ```const rbac = Rbac__factory.connect(rbacModuleAddress, signer);```

**Comment by GabrielKidane96** [comment]
Typechain files shouldn't be pushed as they're generated, also think we can move the location higher up in the hierarchy i.e. have it in the src folder directly / part of types

File: apps/Wallet/src/types/safe-transaction.ts (hunk 1 above)
Comment (GabrielKidane96) [reply apps/Wallet/src/types/safe-transaction.ts]: Maybe good to do cleanup of the owner related code since that wont be necessary anymore with this change

Comment (venoivankovic) [reply apps/Wallet/src/types/safe-transaction.ts]: using single AddressTransactionBody type

**Comment by venoivankovic** [comment]
> Instead of `const moduleContract = new ethers.Contract(allowanceModuleAddress, rbac, signer) as Rbac;` Use typechain as intended `const rbac = Rbac__factory.connect(rbacModuleAddress, signer);`

fixed

**Comment by venoivankovic** [comment]
> Typechain files shouldn't be pushed as they're generated, also think we can move the location higher up in the hierarchy i.e. have it in the src folder directly / part of types

fixed

File: design-decisions.md [hunk 2]
```diff
@@ -15,4 +15,10 @@ Unfortunately, Safe does not provide us with an API to do this on Alfajores. The
 
 As such, for development purposes, we are deploying our contracts on the Ethereum Goerli testnet using this transaction service (https://safe-transaction-goerli.safe.global/). Our final version will then make use of the Celo mainnet supported transaction service API (https://safe-transaction-celo.safe.global/).
 
+### Role Based Access Control Module
```
Comment (venoivankovic) [inline design-decisions.md:18]: @GabrielKidane96

File: apps/Wallet/src/app/api/safe/users/is-user/route.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import { Rbac__factory } from "../../../../../types/typechain/types/config/abis";
+import { getEtherscanSigner } from "../../util/utils";
+
+export async function GET(req: Request, res: Response): Promise<Response> {
+    const body = (await req.json()) as AddressTransactionBody;
+    const signer = getEtherscanSigner(process.env.OWNER_1_PRIVATE_KEY_GOERLI!);
+    const rbacModuleAddress = process.env.RBAC_MODULE_ADDRESS!;
+    const rbac = Rbac__factory.connect(rbacModuleAddress, signer);
+    const safe = process.env.SAFE_ADDRESS!;
+    const isUser = await rbac.isDelegate(safe, body.address)
```
Comment (GabrielKidane96) [inline apps/Wallet/src/app/api/safe/users/is-user/route.ts:11]: This user does not require a private key so lets just pass in the provider instead of signer

File: apps/Wallet/src/app/api/safe/util/utils.ts [hunk 4]
```diff
@@ -3,6 +3,19 @@ import { ethers } from "ethers";
 import SafeApiKit from "@safe-global/api-kit";
 import Safe, { EthersAdapter } from "@safe-global/protocol-kit";
 import { SafeTransactionBody } from "@/types";
+import { etherscanId, network } from "./constants";
+
+export function getSigner() {
+  const provider = new ethers.providers.JsonRpcProvider(process.env.RPC_URL!);
```
Comment (GabrielKidane96) [inline apps/Wallet/src/app/api/safe/util/utils.ts:20]: How do these functions differ from `getEthAdapter`?

File: apps/Wallet/src/types/typechain/abis/rbac.json [hunk 5]
```diff
@@ -0,0 +1 @@
...internalType":"uint48","name":"","type":"uint48"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"contract GnosisSafe","name":"safe","type":"address"},{"internalType":"address payable","name":"to","type":"address"},{"internalType":"uint96","name":"amount","type":"uint96"}],"name":"executeTransfer","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"safe","type":"address"},{"internalType":"uint48","name":"start","type":"uint48"},{"internalType":"uint8","name":"pageSize","type":"uint8"}],"name":"getDelegates","outputs":[{"internalType":"address[]","name":"results","type":"address[]"},{"internalType":"uint48","name":"next","type":"uint48"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"contract GnosisSafe","name":"safe","type":"address"},{"internalType":"address","name":"delegate","type":"address"}],"name":"isDelegate","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"delegate","type":"address"}],"name":"removeDelegate","outputs":[],"stateMutability":"nonpayable","type":"function"}]
```
Comment (GabrielKidane96) [inline apps/Wallet/src/types/typechain/abis/rbac.json:1]: remove

Comment (venoivankovic) [reply apps/Wallet/src/types/typechain/abis/rbac.json:1]: you cant`
    "typechain": "typechain --target ethers-v5 --out-dir ./src/types/typechain/types/config/abis './src/types/typechain/abis/*.json'"`
the abi is used as an input to generate the typings

File: apps/Wallet/src/app/api/safe/util/utils.ts (hunk 4 above)
Comment (venoivankovic) [reply apps/Wallet/src/app/api/safe/util/utils.ts:20]: i want the signer, not the ethadapter (different objects are returned)
etherscansigner is added because using default signer was giving back cannot estimate gas errors, etherscansigner was not

File: apps/Wallet/src/app/api/safe/users/is-user/route.ts (hunk 3 above)
Comment (venoivankovic) [reply apps/Wallet/src/app/api/safe/users/is-user/route.ts:11]: fixed

File: apps/Wallet/src/app/api/safe/util/utils.ts (hunk 4 above)
Comment (GabrielKidane96) [reply apps/Wallet/src/app/api/safe/util/utils.ts:20]: You can simply get the signer from the adapter by calling `.getSigner()`. Regarding etherscan should we not then just set it as the default instead of having two functions with different signers? I'd suggest keeping only the `getEthAdapter` adding the etherscan there and only using that to get the required setup for blockchain calls, apart from that we should probably set the etherscanId in the .env

File: apps/Wallet/src/app/api/safe/users/is-user/route.ts (hunk 3 above)
Comment (GabrielKidane96) [reply apps/Wallet/src/app/api/safe/users/is-user/route.ts:11]: same here, can just use getEthAdapter and use the `.getProvider` function

File: apps/Wallet/src/app/api/safe/util/utils.ts (hunk 4 above)
Comment (venoivankovic) [reply apps/Wallet/src/app/api/safe/util/utils.ts:20]: EthersAdapter comes from the Safe protocol kit, why do you want to use it just to create a signer, please explain.
if you use .getSigner then you have to check that its defined every single time you want it back, and ur creating a Safe protocol kit object, why exactly? 
creating a signer requires only calling ethers, a more common library. if you want to do that it will look like this, complicated and convoluted imo, as well as the safe protocol kit dependency
```
async function getRemoveUserCallData(rbacModuleAddress: string, delegateAddress: string) {
  const ethAdapter = getEthAdapter(process.env.OWNER_1_PRIVATE_KEY_GOERLI!);
  const signer = ethAdapter.getSigner();
  if(!signer) {
    throw new Error("no signer")
  }
  const rbac = Rbac__factory.connect(rbacModuleAddress, signer);
  const tx = await rbac.populateTransaction.removeDelegate(delegateAddress);
  return tx.data;
}
```


id really prefer not to create objects using the safe protocol kit unless we need to
but ok, ive removed the getSigner method and refactored so getEthAdapter uses getEtherscanSigner and made it an env var
`ETHERSCAN_ID="9QA7YFDY8APXCMB32IU6KUEA6F83HH6VPZ"`

Comment (GabrielKidane96) [reply apps/Wallet/src/app/api/safe/util/utils.ts:20]: What I find convoluted is having 3 functions that can all do exactly the same thing, there is no difference in the signer that comes from the adapter or the signer that comes from calling getSigner, the way I see it, as a developer, having one object that provides you with the necessities to interact with the blockchain is better than having 3 functions that can all do it.

Comment (GabrielKidane96) [reply apps/Wallet/src/app/api/safe/util/utils.ts:20]: I'll leave it up to you

**Comment by GabrielKidane96** [comment]
We'd technically want to remove the usage of `OWNER_1_PRIVATE_KEY_GOERLI` and pass in the pk to the functions that make use of user management since this will be something that the admin will be in control of.

File: apps/Wallet/src/app/api/safe/util/utils.ts (hunk 4 above)
Comment (venoivankovic) [reply apps/Wallet/src/app/api/safe/util/utils.ts:20]: ok, in utils theres a getEtherscanProvider and getSigner function now, which im exporting in the routes when we need signers and providers
these functions are also called by getEthAdapter
i think this is the best way to do it

**Comment by venoivankovic** [comment]
> We'd technically want to remove the usage of `OWNER_1_PRIVATE_KEY_GOERLI` and pass in the pk to the functions that make use of user management since this will be something that the admin will be in control of.

cool, in the add user and remove user routes we now use the pk of the admin thats logged in

**Review by GabrielKidane96 — APPROVED** [review]
